### ISSUE TITLE
Consider adding Windows plugin API to post talks to main thread

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#79213]
There are plugin APIs that can only be called from the main thread, but there's not a particularly good way for Windows plugins to move calls to that thread AFAIK. Options include:
- Each plugin doing the hidden-window message loop implementation themselves.
- Using the top-level window proc override registration to send themselves a custom message and hope nobody else uses it in a way that interferes.

This came up in the Texture PR and the solution there was to make most of it thread-safe; applying that more broadly is another option (e.g., to event streams). But a generic way to move things to the main thread would avoid needing to make lots of thread-safe APIs.

/cc @cbracken

**Comment by Sunbreak** [comment]
> * Each plugin doing the hidden-window message loop implementation themselves.

hidden-window message loop implementation for further reference: https://github.com/flutter/engine/pull/24232

**Comment by sebastianhaberey** [comment]
Just dropping in to say this would be immensely helpful for an [issue](https://github.com/sebastianhaberey/nsd/issues/46) I am currently facing with my plugin. Until then, would you advise me to use the linked commit as a template for a hidden message loop within the plugin?

**Comment by stuartmorgan** [comment]
> Just dropping in to say this would be immensely helpful for an [issue](https://github.com/sebastianhaberey/nsd/issues/46) I am currently facing with my plugin.

As discussed in https://github.com/flutter/engine/pull/31134, the preferred solution to the problem you describe in that issue is good the Windows embedding API to be made thread safe.

**Comment by sebastianhaberey** [comment]
From the issue you linked it seems that this is still in the planning stage. Is there a temporary solution you'd recommend until the feature is part of Flutter? Or should we just wait?

**Comment by stuartmorgan** [comment]
> From the issue you linked it seems that this is still in the planning stage.

Yes, but so is this, and the current thinking is that we will very likely take the other approach, and close this.

> Is there a temporary solution you'd recommend until the feature is part of Flutter?

Until that's implemented the solution is to send messages from the platform thread, as the API currently requires.

> Or should we just wait?

If you have a plugin that is currently crashing clients because it's violating the API requirements I wouldn't suggest waiting, but it's your choice.

**Comment by stuartmorgan** [comment]
It looks like that discussion never got captured in an issue, so I've filed https://github.com/flutter/flutter/issues/134346

**Comment by loic-sharma** [comment]
On https://github.com/flutter/flutter/issues/134346#issuecomment-2035157295 I proposed a strawman API for this issue:

```cpp
class FlutterEngine : public PluginRegistry {

  // ...

  // Schedule a callback to be called on the platform thread.
  //
  // This may be called on any thread. The callback is executed only
  // once on the platform thread.
  void PostPlatformThreadTask(std::function<void()> callback);
}
```

```cpp
// Schedule a callback to be invoked on the engine's platform thread.
//
// This may be called on any thread. The callback is executed only
// once on the platform thread.
FLUTTER_EXPORT void FlutterDesktopEnginePostPlatformThreadTask(
    FlutterDesktopEngineRef engine,
    VoidCallback callback,
    void* user_data);
```

**Comment by Ae-Mc** [comment]
+1

**Comment by loic-sharma** [comment]
Added by: https://github.com/flutter/flutter/pull/187365
